fn: generalize type of t in UnwrapOrFail

starius opened the pull request:
## Change Description

It is needed to pass other types satisfying needed interfaces to methods `Option.UnwrapOrFail` and `Result.UnwrapOrFail`.

An example of such a type is `*lntest.HarnessTest`. It embeds `*testing.T` and has all the methods but can't be passed directly as `*testing.T`, but can be passed now as an instance of `Testing` interface.

## Steps to Test

This is an internal change needed to simplify development.

## Pull Request Checklist
### Testing
- [ ] Your PR passes all CI checks.
- [ ] Tests covering the positive and negative (error paths) are included.
- [ ] Bug fixes contain tests triggering the bug to prevent regressions.

### Code Style and Documentation
- [ ] The change obeys the [Code Documentation and Commenting](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md#CodeDocumentation) guidelines, and lines wrap at 80.
- [ ] Commits follow the [Ideal Git Commit Structure](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md#IdealGitCommitStructure).
- [ ] Any new logging statements use an appropriate subsystem and logging level.
- [ ] Any new lncli commands have appropriate tags in the comments for the rpc in the proto file.
- [ ] [There is a change description in the release notes](https://github.com/lightningnetwork/lnd/tree/master/docs/release-notes), or `[skip ci]` in the commit message for small changes.

📝 Please see our [Contribution Guidelines](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md) for further guidance.

ProofOfKeags commented:
As stated in #9141 review, this is a NACK from me. If we do this we should tightly define the testing interface we want to conform to and then add a separate file that defines it and its expected properties, rather than putting it in option.go or other specific type files.

starius commented:
> As stated in #9141 review, this is a NACK from me. If we do this we should tightly define the testing interface we want to conform to and then add a separate file that defines it and its expected properties, rather than putting it in option.go or other specific type files.

I moved the interface to a separate file (testing.go) and separated it to multiple parts. The base interface (TestingT) has only method `Errorf`. This is needed to support [assert.CollectT](https://pkg.go.dev/github.com/stretchr/testify@v1.9.0/assert#CollectT) as well, which can appear in `require.EventuallyWithT` callback. "require" package itself defines [assert.TestingT](https://pkg.go.dev/github.com/stretchr/testify@v1.9.0/assert#TestingT) which has only one method: `Errorf`.

I added a test for the method testing it with real `t` and with a mock to make sure `Helper()` and `FailNow()` are called.

starius commented:
I found a bug in `Result.UnwrapOrFail` implementation. It used to return zero value in case of success. It should return the value stored. I put it into a separate commit "fn: fix UnwrapOrFail success case",

ProofOfKeags commented:
> I found a bug in Result.UnwrapOrFail implementation. It used to return zero value in case of success. It should return the value stored. I put it into a separate commit "fn: fix UnwrapOrFail success case",

Is this still true, I remember finding the same bug and fixed it, I thought.
